src/TAMSEquationSystem.C

```diff
@@ -0,0 +1,615 @@
+    const auto& U = *meta_data.get_field<VectorFieldType>(stk::topology::NODE_RANK, "velocity");
+    field_copy(realm_.meta_data(), realm_.bulk_data(), U, avgU, realm_.get_activate_aura());
+
+    // Copy dudx to average dudx
+    GenericFieldType &avgDudx = avgDudx_->field_of_state(stk::mesh::StateNP1);
+    const auto& dudx = *meta_data.get_field<GenericFieldType>(stk::topology::NODE_RANK, "dudx");
+    field_copy(realm_.meta_data(), realm_.bulk_data(), dudx, avgDudx, realm_.get_activate_aura());
+
+    ScalarFieldType* turbKinEne_ = meta_data.get_field<ScalarFieldType>(
+      stk::topology::NODE_RANK, "turbulent_ke");
+    ScalarFieldType* tvisc_ = meta_data.get_field<ScalarFieldType>(
+      stk::topology::NODE_RANK, "turbulent_viscosity");
+    ScalarFieldType& tkeNp1 = turbKinEne_->field_of_state(stk::mesh::StateNP1);
+
+    // define some common selectors
+    stk::mesh::Selector s_all_nodes =
+      (meta_data.locally_owned_part() | meta_data.globally_shared_part()) &
+      stk::mesh::selectField(*avgDudx_);
+
+    stk::mesh::BucketVector const& buckets =
+      realm_.get_buckets(stk::topology::NODE_RANK, s_all_nodes);
+    for (stk::mesh::BucketVector::const_iterator ib = buckets.begin();
+         ib != buckets.end(); ++ib) {
+      stk::mesh::Bucket& b = **ib;
+      const stk::mesh::Bucket::size_type length = b.size();
+
+      double* tke = stk::mesh::field_data(tkeNp1, b);
+      double* tvisc = stk::mesh::field_data(*tvisc_, b);
+      double* avgProd = stk::mesh::field_data(*avgProduction_, b);
+      double* rho = stk::mesh::field_data(*avgDensity_, b);
+
+      for (stk::mesh::Bucket::size_type k = 0; k < length; ++k) {
+        // Initialize average production to mean production
+        const double* avgDudx = stk::mesh::field_data(*avgDudx_, b[k]);
+        double* tij = new double[nDim * nDim];
+        for (int i = 0; i < nDim; ++i) {
+          for (int j = 0; j < nDim; ++j) {
+            const double avgSij =
+              0.5 * (avgDudx[i * nDim + j] + avgDudx[j * nDim + i]);
+            tij[i * nDim + j] = 2.0 * tvisc[k] * avgSij;
+          }
+          tij[i * nDim + i] -= 2.0 / 3.0 * rho[k] * tke[k];
+        }
+
+        double* Pij = new double[nDim * nDim];
+        for (int i = 0; i < nDim; ++i) {
+          for (int j = 0; j < nDim; ++j) {
+            Pij[i * nDim + j] = 0.0;
+            for (int m = 0; m < nDim; ++m) {
+              Pij[i * nDim + j] += avgDudx[i * nDim + m] * tij[j * nDim + m] +
+                                   avgDudx[j * nDim + m] * tij[i * nDim + m];
+            }
+            Pij[i * nDim + j] *= 0.5;
+          }
+        }
+
+        double instProd = 0.0;
+        for (int i = 0; i < nDim; ++i)
+          instProd += Pij[i * nDim + i];
+
+        avgProd[k] = instProd;
+
+        delete[] tij;
+        delete[] Pij;
+      }
+    }
+  }
+
+  compute_averages();
+
+  // FIXME: Moved this to SST Eqn Systems for now since mdot has not 
```

Comment:
We could depend on the requirement that LowMach always goes first in Nalu-Wind, and therefore do the mdot calculation in `initial_work` of `LowMach`?

Reply:
Just to make sure I understand what you're saying, you are suggesting moving the `averageMdot` stuff to `LowMach` within a `turbulence model == SST_TAMS` statement?